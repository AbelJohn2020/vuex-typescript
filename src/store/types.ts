export type user = {
    id: number,
    fieldName: {
        name: string,
        validName: string,
    },
    fieldLastname: {
        lastname: string,
        validLastname: string,
    },
    fieldNationality: {
        nationality: string,
        validNationality: string,
    },
    fieldIdentification: {
        identification: string,
        validIdentification: string,
    },
    fieldDocument: {
        document: string,
        validDocument: string,
    },
};

export type state = {
    user: user,
    letters: boolean,
    continue: string,
    language: string,
    users: user[] | [],
    regNumbers: RegExp,
    regex: RegExp,
    regNationality: RegExp,
    maxTickets: number[],
    tickets: number,
    validMultiForm: boolean[] | [],
}