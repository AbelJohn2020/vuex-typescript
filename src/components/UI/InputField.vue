<template>
  <div class="form-control" :class="{invalid: validation}">
    <label :for="forLabel" id="for-label">
      {{ forLabel }}
      <slot name="valid"></slot>
    </label>
    <slot></slot>
    <p v-if="validation">invalid field</p>
  </div>
</template>

<script>
export default {
    props: [
        'forLabel',
        'validation',
    ]
}
</script>

<style>
.form-control {
  margin: 1rem 0;
  width: 100%;
  display: flex;
  flex-direction: column;
}

#for-label {
  display: flex;
}

label {
    text-transform: capitalize;
}

label, p {
  box-sizing: border-box;
  font-weight: bold;
  text-align: start;
}

p {
    text-transform: capitalize;
    padding: 0;
    margin: 0.4rem 0 0 0;
}

.form-control.invalid label,
.form-control.invalid p {
  color: red;
}
</style>