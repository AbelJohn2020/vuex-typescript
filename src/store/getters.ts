import { state } from './types';

export default {
    getTickets(state: state) {
        return Array.from({length: state.tickets}, (_, i) => i + 1)
    },
    enableSubmit(state: state) {
        // return state.validMultiForm.includes(false) ? true : false;
        const filterFalse = state.validMultiForm.filter(element => element === false);
        return filterFalse.length > 0 ? true : false;
    },
}