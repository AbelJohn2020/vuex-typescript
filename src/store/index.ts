import { createStore } from 'vuex';
import rootGetters from './getters';
import rootMutations from './mutations';

const store = createStore({
    
    state() {
        return {
            user: {
                id: 0,
                fieldName: {
                    name: '',
                    validName: 'pending',
                },
                fieldLastname: {
                    lastname: '',
                    validLastname: 'pending',
                },
                fieldNationality: {
                    nationality: '',
                    validNationality: 'pending',
                },
                fieldIdentification: {
                    identification: '',
                    validIdentification: 'pending',
                },
                fieldDocument: {
                    document: '',
                    validDocument: 'pending',
                },
            },

            letters: false,
            continue: 'form',
            language: 'english',
            users: [],
            validMultiForm: [],
            
            maxTickets: Array.from({length: 4}, (_, i) => i + 1),
            tickets: 0,
            regNumbers: /^[0-9]+$/,
            regex: /^[a-zñ A-ZÑáéíóúÁÉÍÓÚ'.]*$/,
            regNationality: /^[a-zñA-ZÑ]*$/,
        };
    },
    mutations: rootMutations,

    getters: rootGetters,
});

export default store;