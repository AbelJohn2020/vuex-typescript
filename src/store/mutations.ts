import { state, user } from './types';

export default {
    handleChangeSelect (state: state, event: Event) {
        state.tickets = parseInt((event.target as HTMLInputElement).value);
    },
    handleSelect(state: state, getTickets: number[]) {
        let ticket: string;
        state.users = [];
        const users: user[] = state.users;
  
        for(ticket in getTickets) {
            users.push({...state.user, id: parseInt(ticket)})
        }
        return users;
    },

    handleBack(state: state) {
        state.continue = 'form';
    },

    handleAccept(state: state) {
        state.continue = 'form';
    },

    handleNext(state: state) {
        state.continue = 'review';
    },

    getTickets(state: state) {
        Array.from({length: state.tickets}, (_, i) => i + 1)
    },

    multiForm(state: state, tickets: number) {
        let i;
        state.validMultiForm = [];

        for(i= 0; i < tickets; i++) {
            state.validMultiForm = [...state.validMultiForm, false]
        }
        return state.validMultiForm;
    },

    handleInName(state:state, name: {idx: number, event: Event}) {
        const { idx, event } = name
        state.users[idx] = {...state.users[idx], fieldName: {...state.users[idx].fieldName, name: (event.target as HTMLInputElement).value}};
    },

    handleInLastname(state:state, lastname: {idx: number, event: Event}) {
        const { idx, event } = lastname;
        state.users[idx] = {...state.users[idx], fieldLastname: {...state.users[idx].fieldLastname, lastname: (event.target as HTMLInputElement).value}};
    },

    handleInNationality(state:state, nationality: {idx: number, event: Event}) {
        const { idx, event } = nationality;
        state.users[idx] = { ...state.users[idx], fieldNationality: {...state.users[idx].fieldNationality, nationality: (event.target as HTMLInputElement).value}};
    },

    typeIDField(state: state, identification: {event: Event, idx: number}) {
        const { event, idx } = identification;

        state.users[idx] = { ...state.users[idx], fieldIdentification: { ...state.users[idx].fieldIdentification, identification: (event.target as HTMLInputElement).value }};
        const idType = state.users[idx].fieldIdentification.identification;

        if(idType === '') {
            state.users[idx] = { ...state.users[idx], fieldIdentification: { ...state.users[idx].fieldIdentification, validIdentification: 'invalid' }};
        } else {
            state.users[idx] = { ...state.users[idx], fieldIdentification: { ...state.users[idx].fieldIdentification, validIdentification: 'valid' }};
            state.users[idx] = { ...state.users[idx], fieldDocument: { ...state.users[idx].fieldDocument, document: '' }};
            state.users[idx] = { ...state.users[idx], fieldDocument: { ...state.users[idx].fieldDocument, validDocument: 'pending' }};
        }
    },

    documentField(state: state, document: {event: Event, idx: number}) {
        const { event, idx } = document;

        state.users[idx] = { ...state.users[idx], fieldDocument: { ...state.users[idx].fieldDocument, document: (event.target as HTMLInputElement).value }}
        const documentValue = state.users[idx].fieldDocument.document;
        const validDcument = state.regNumbers.test(documentValue);
        const len = documentValue.length;
        const doclen = ( state.users[idx].fieldIdentification.identification === 'dni' ) ? 8 : 9

        if((len === doclen) && validDcument) {
            state.users[idx] = { ...state.users[idx], fieldDocument: { ...state.users[idx].fieldDocument, validDocument: 'valid' }};
        } else {
            state.users[idx] = { ...state.users[idx], fieldDocument: { ...state.users[idx].fieldDocument, validDocument: 'invalid' }};
        }
    },

    documentError(state: state, idx: number) {
        const documentValue = state.users[idx].fieldDocument.document;
        const validDocument = state.regNumbers.test(documentValue);

        if(validDocument) {
            state.letters = false;
        } else {
            state.letters = true;
        }
    },

    onBlurName(state: state, idx: number) {
        const name = state.users[idx].fieldName.name;
        const validName = state.regex.test(name);
        
        if(name.length > 1 && validName) {
            state.users[idx] = {...state.users[idx], fieldName: {...state.users[idx].fieldName, validName: 'valid'}}
        } else {
            state.users[idx] = {...state.users[idx], fieldName: {...state.users[idx].fieldName, validName: 'invalid'}}
        }
    },

    onBlurLastname(state: state, idx: number) {
        const lastname = state.users[idx].fieldLastname.lastname;
        const validLastname = state.regex.test(lastname);
        
        if(lastname.length > 1 && validLastname) {
            state.users[idx] = {...state.users[idx], fieldLastname: {...state.users[idx].fieldLastname, validLastname: 'valid'}}
        } else {
            state.users[idx] = {...state.users[idx], fieldLastname: {...state.users[idx].fieldLastname, validLastname: 'invalid'}}
        }
    },

    onBlurNationality(state: state, idx: number) {
        const nationality = state.users[idx].fieldNationality.nationality;
        const validNationality = state.regNationality.test(nationality);
        
        if(nationality.length > 1 && validNationality) {
            state.users[idx] = {...state.users[idx], fieldNationality: {...state.users[idx].fieldNationality, validNationality: 'valid'}}
        } else {
            state.users[idx] = {...state.users[idx], fieldNationality: {...state.users[idx].fieldNationality, validNationality: 'invalid'}}
        }
    },

    onBlurIdentification(state: state, idx: number) {
        const identification = state.users[idx].fieldIdentification.identification;
        if(identification.length > 1) {
            state.users[idx] = {...state.users[idx], fieldIdentification: {...state.users[idx].fieldIdentification, validIdentification: 'valid'}}
        } else {
            state.users[idx] = {...state.users[idx], fieldIdentification: {...state.users[idx].fieldIdentification, validIdentification: 'invalid'}}
        }
    },

    validMultipleForm(state: state, idx: number) {
        const validName = state.users[idx].fieldName.validName;
        const validLastname = state.users[idx].fieldLastname.validLastname;
        const validNationality = state.users[idx].fieldNationality.validNationality;
        const validIdentification = state.users[idx].fieldIdentification.validIdentification;
        const validDocument = state.users[idx].fieldDocument.validDocument;

        if(validName === 'valid' && validLastname === 'valid' && validNationality === 'valid' && validIdentification === 'valid' && validDocument === 'valid') {
            state.validMultiForm[idx] = true;
        } else {
            state.validMultiForm[idx] = false;
        }
    },

    handleSubmit(state: state) {
        state.continue= 'finish';

        state.user= {
            id: 0,
            fieldName: {
                name: '',
                validName: 'pending',
            },
            fieldLastname: {
                lastname: '',
                validLastname: 'pending',
            },
            fieldNationality: {
                nationality: '',
                validNationality: 'pending',
            },
            fieldIdentification: {
                identification: '',
                validIdentification: 'pending',
            },
            fieldDocument: {
                document: '',
                validDocument: 'pending',
            },
        };

        state.language= 'english';
        state.letters= false;
        state.users= [];
        state.validMultiForm= [];
        
        state.tickets= 0;
    },
}