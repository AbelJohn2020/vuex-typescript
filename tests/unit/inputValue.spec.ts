import store from '@/store';
import { mount } from '@vue/test-utils';
import InputField from '../../src/components/UI/InputField.vue';

describe('Input field Name', () => { 
    const wrapper = mount(InputField, {
        slots: {
            default: `
                <input placeholder="Name" id="name" type="text" v-model="valueName"/>
            `,
        },
        props: {
            forLabel: 'name',
            validation: false
        },
        data() {
            return {
                valueName: '',
            };
        },
    });

    const inputValue = wrapper.find('input[type="text"]');
    const regex = store.state.regex;


    test('The default Slot in InputField will be Abel Antonio', async () => { 

        await inputValue.setValue('Abel Antonio');

        expect((inputValue.element as HTMLInputElement).value).toBe('Abel Antonio');
    });

    test('Should return Invalid Field', async () => { 

        await inputValue.setValue('Abel Antonio11');

        const isNotValid = mount(InputField, {
            ...wrapper, 
            slots:{ default: '<input placeholder="Name" id="name" type="text" v-model="valueName"/>' },
            props: {...wrapper.props(), validation: true}
        });

        const notValid = regex.test((inputValue.element as HTMLInputElement).value);

        expect(notValid).toBe(false);
        expect(isNotValid.find('p').element.textContent).toContain('invalid field');
        expect(isNotValid.html()).toMatchSnapshot();
     });

    test('should return Valid Field', async () => { 
        await inputValue.setValue('Abel John');
        const valid = regex.test((inputValue.element as HTMLInputElement).value);
        const test = mount(InputField, {
            ...wrapper, 
            slots:{ default: '<input placeholder="Name" id="name" type="text" v-model="valueName"/>', valid: '<p>valid field</p>' }, 
            props: {...wrapper.props()}
        });

        expect(valid).toBe(true);
        expect(test.find('p').element.textContent).toContain('valid fiel');
        expect(test.html()).toMatchSnapshot();
    });
 });


describe('Input Field Nationality', () => { 
    const wrapper = mount(InputField, {
        slots: {
            default: `
                <input placeholder="Name" id="name" type="text" v-model="valueNationality"/>
            `,
        },
        props: {
            forLabel: 'nationality',
            validation: false
        },
        data() {
            return {
                valueNationality: '',
            };
        },
    });

    const inputValue = wrapper.find('input[type="text"]');
    const regex = store.state.regNationality;


    test('The default Slot in InputField will be Peruano', async () => { 

        await inputValue.setValue('Peruano');
        expect((inputValue.element as HTMLInputElement).value).toBe('Peruano');
    });

    test('Should return Invalid Field', async () => { 

        await inputValue.setValue("Peruano'");

        const isNotValid = mount(InputField, {
            ...wrapper, 
            slots:{ default: '<input placeholder="Name" id="name" type="text" v-model="valueNationality"/>' },
            props: {...wrapper.props(), validation: true}
        });

        const notValid = regex.test((inputValue.element as HTMLInputElement).value);

        expect(notValid).toBe(false);
        expect(isNotValid.find('p').element.textContent).toContain('invalid field');
        expect(isNotValid.html()).toMatchSnapshot();
     });

    test('should return Valid Field', async () => { 
        await inputValue.setValue('peruano');
        const valid = regex.test((inputValue.element as HTMLInputElement).value);
        const test = mount(InputField, {
            ...wrapper, 
            slots:{ default: '<input placeholder="Name" id="name" type="text" v-model="valueNationality"/>', valid: '<p>valid field</p>' }, 
            props: {...wrapper.props()}
        });

        expect(valid).toBe(true);
        expect(test.find('p').element.textContent).toContain('valid fiel');
        expect(test.html()).toMatchSnapshot();
    });
  })

  describe('Input Field Identification', () => { 
    const wrapper = mount(InputField, {
        slots: {
            default: `
                <input placeholder="Name" id="name" type="text" v-model="valueIdentification"/>
            `,
        },
        props: {
            forLabel: 'ID',
            validation: false
        },
        data() {
            return {
                valueIdentification: '',
            };
        },
    });

    const inputValue = wrapper.find('input[type="text"]');
    const regex = store.state.regNumbers;


    test('The default Slot in InputField will be Peruano', async () => { 

        await inputValue.setValue('123456789');
        expect((inputValue.element as HTMLInputElement).value).toBe('123456789');
    });

    test('Should return Invalid Field', async () => { 

        await inputValue.setValue("12345678d");

        const isNotValid = mount(InputField, {
            ...wrapper, 
            slots:{ default: '<input placeholder="Name" id="name" type="text" v-model="valueIdentification"/>' },
            props: {...wrapper.props(), validation: true}
        });

        const notValid = regex.test((inputValue.element as HTMLInputElement).value);

        expect(notValid).toBe(false);
        expect(isNotValid.find('p').element.textContent).toContain('invalid field');
        expect(isNotValid.html()).toMatchSnapshot();
     });

    test('should return Valid Field', async () => { 
        await inputValue.setValue('123456789');
        const valid = regex.test((inputValue.element as HTMLInputElement).value);
        const test = mount(InputField, {
            ...wrapper, 
            slots:{ default: '<input placeholder="Name" id="name" type="text" v-model="valueIdentification"/>', valid: '<p>valid field</p>' }, 
            props: {...wrapper.props()}
        });

        expect(valid).toBe(true);
        expect(test.find('p').element.textContent).toContain('valid fiel');
        expect(test.html()).toMatchSnapshot();
    });
  })