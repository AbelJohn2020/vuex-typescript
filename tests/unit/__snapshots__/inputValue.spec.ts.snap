// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Field Identification Should return Invalid Field 1`] = `
<div class="form-control invalid"><label for="ID" id="for-label">ID </label><input placeholder="Name" id="name" type="text">
  <p>invalid field</p>
</div>
`;

exports[`Input Field Identification should return Valid Field 1`] = `
<div class="form-control"><label for="ID" id="for-label">ID <p>valid field</p></label><input placeholder="Name" id="name" type="text">
  <!--v-if-->
</div>
`;

exports[`Input Field Nationality Should return Invalid Field 1`] = `
<div class="form-control invalid"><label for="nationality" id="for-label">nationality </label><input placeholder="Name" id="name" type="text">
  <p>invalid field</p>
</div>
`;

exports[`Input Field Nationality should return Valid Field 1`] = `
<div class="form-control"><label for="nationality" id="for-label">nationality <p>valid field</p></label><input placeholder="Name" id="name" type="text">
  <!--v-if-->
</div>
`;

exports[`Input field Name Should return Invalid Field 1`] = `
<div class="form-control invalid"><label for="name" id="for-label">name </label><input placeholder="Name" id="name" type="text">
  <p>invalid field</p>
</div>
`;

exports[`Input field Name should return Valid Field 1`] = `
<div class="form-control"><label for="name" id="for-label">name <p>valid field</p></label><input placeholder="Name" id="name" type="text">
  <!--v-if-->
</div>
`;
